### DISCUSSION TITLE
`setValue` without triggering change events

### DISCUSSION THREAD

**Discussion by den1k** [microsoft/monaco-editor#3628] [Q&A]
We only want to listen to user-triggered events. It looks like `_resetValue` would allow for this (I get it, it's supposed to be private). However, we can't seem to use it because the necessary function `toRawText` defined on `TextModel` seems to be unaccessible.

How can we change the editor text value without triggering a change event?

**Comment by alexdima** [comment]
I'm not sure I understand the intent here. Any changes done to the model will result in change events being emitted (e.g. to update the view or the language services that provide rich language features). What would be the use case for changing a model and inhibiting events?

**Comment by den1k** [comment]
The distinction lies in the event source. Are events triggered by the user or code? In my case I want to listen to user triggered events, but also set the editor content and cursor position using code. The current solution is to unbind and rebind listeners each time. Is this clearer?

**Comment by alexdima** [comment]
This is a common problem I also hit often. Usually I use a code pattern like the following:

```js
var ignoreEventBecauseIAmTriggeringIt = false;

emitter.onBla(function(e) {
  if (ignoreEventBecauseIAmTriggeringIt) {
    return;
  }
  console.log('Now I should react to ', e, ' because someone else caused it');
});

...

// Do something that makes `onBla` fire:
try {
  ignoreEventBecauseIAmTriggeringIt = true;
  emitter.changeBla(..);
} finally {
  ignoreEventBecauseIAmTriggeringIt = false;
}
```

**Comment by den1k** [comment]
Thats essentially what I'm doing @alexandrudima. To avoid race conditions I think it would be great to have a way to **know** what caused the event, e.g. the user or the code. This would also be valuable for widgets and other feedback.

**Comment by chinchang** [comment]
+1. For eg. In Codemirror for every change you also get it's origin - if it was a user input or setValue call.

**Comment by artparks** [comment]
agreed it would be useful to have this distinction - would make it play with angular's ControlValueAccessor API much better: https://github.com/angular/angular/issues/20230#issuecomment-361799387

**Comment by maxxcs** [comment]
It would be great if the 'source' parameter of the 'executeEdits' function could be obtained in the 'onDidChangeModelContent' function.

**Comment by manideepmothe641** [comment]
after toggling to a different file , im unable to get previous changes using ctrl + z for previous file..

any help would be appreciable

**Comment by stanchovy** [comment]
This is still an issue and a fairly important one. Any update on this?

**Comment by worksofliam** [comment]
This has cropped up for me as well. Trying to build an instance so an instance of the editor and model can be edited from a web socket event, but I don't want it to trigger the event

**Comment by stanchovy** [comment]
I'll just add here that it's fairly standard practice to be have this functionality.... Ace editor, Code mirror, even back as far as Jquery widgets have this.

**Comment by Sharkou** [comment]
I had the same problem and found a solution/workaround : isFlush property from onDidChangeModelContent event. Try this :
```js
editor.onDidChangeModelContent(e => {
  if (e.isFlush) {
    // true if setValue call
  } else {
    // false if user input
  }
});

**Reply by stephengeorge568** [reply]
isFlush is only true when the model gets reset... i.e editor.setValue(). Less than ideal and maybe not a valid solution for some use cases. I can't imagine a good reason that a source is not apart of a Change event metadata. Its included as an argument to executeEdits() but evidently that source is not used anywhere.

**Comment by shaneapen** [comment]
Any update on this issue?

**Comment by InfinityDJKing** [comment]
Hi, I have same problem.
I want to use websocket to sync every code change, but is will trigger a dead loop

**Comment by InfinityDJKing** [comment]
> I'm not sure I understand the intent here. Any changes done to the model will result in change events being emitted (e.g. to update the view or the language services that provide rich language features). What would be the use case for changing a model and inhibiting events?

Hi, I have same problem.
I want to use websocket to sync every code change, but is will trigger a dead loop.
Can I get which source did it come from like onDidChangeCursorSelection?(ICursorSelectionChangedEvent)

**Comment by softmarshmallow** [comment]
I need this to prevent event if the change is made by formatter.
`e.isFlush` seems not the solution to me. :(

**Comment by halabooda** [comment]
@DanielJefferson1990 

```
editor. onDidChangeCursorSelection(ev => {
if (ev.reason === CursorChangeReason.ContentFlush) {
    // true if setValue call
  } else {
    // false if user input
  }
}
```

**Comment by lafkpages** [comment]
Any updates on this?

**Comment by chandrakantmkarale** [comment]
Facing similar issue; the content change through the code is triggering the monaco.editor.IModelContentChangedEvent.   And there is no way it can be ignored;   there is no parameter to differentiate between keyboard/user-interaction originated event and code update generated event.
Something of sort - https://github.com/suren-atoyan/monaco-react/issues/91
Need this support.

**Comment by kinthaiofficial** [comment]
Setting Monaco's value without triggering change events is a common need when you want to apply programmatic changes (e.g., loading initial content, applying remote edits) without triggering your own change handlers.

**The cleanest approach: use `pushEditOperations`**:
```typescript
const model = editor.getModel()!;

// This doesn't trigger 'onDidChangeContent' in the same way as setValue
// But it still fires change events — this is intentional in Monaco

// Workaround: flag-based suppression
let ignoreChanges = false;

editor.onDidChangeModelContent(() => {
  if (ignoreChanges) return;
  // your normal change handler
  handleChange(editor.getValue());
});

// When you want to set value without triggering handler:
ignoreChanges = true;
editor.setValue(newContent);
// NOTE: setValue also resets undo history!
ignoreChanges = false;
```

**Better approach: use `applyEdits` + `executeEdits`**:
```typescript
// If you know the diff (from remote change):
const model = editor.getModel()!;
model.applyEdits([{
  range: model.getFullModelRange(),
  text: newContent,
}]);
// This fires onDidChangeContent but preserves undo stack
```

**If you need to preserve undo history AND suppress events**:
```typescript
// Decorations approach: apply without changing model content
// (Only works if you're highlighting, not changing text)

// For actual content changes without history:
model.setValue(newContent);  // Clears undo history (intentional for "loading" scenario)
```

**Practical patterns**:
- Initial load: use `setValue` (clears history — correct for fresh load)
- Remote collaborative changes: use `applyEdits` (preserves local undo stack)
- Internal reformatting: flag-based suppression as shown above

What's your use case — initial load, collaborative editing, or something else?
